compiler/rustc_infer/src/infer/nll_relate/mod.rs

```diff
@@ -603,61 +603,78 @@ where
-            // to correctly handle cases like
-            //
-            //     for<'a> fn(&'a u32, &'a u32) == for<'b, 'c> fn(&'b u32, &'c u32)
-            //
-            // Somewhat surprisingly, these two types are actually
-            // **equal**, even though the one on the right looks more
-            // polymorphic. The reason is due to subtyping. To see it,
-            // consider that each function can call the other:
-            //
-            // - The left function can call the right with `'b` and
-            //   `'c` both equal to `'a`
-            //
-            // - The right function can call the left with `'a` set to
-            //   `{P}`, where P is the point in the CFG where the call
-            //   itself occurs. Note that `'b` and `'c` must both
-            //   include P. At the point, the call works because of
-            //   subtyping (i.e., `&'b u32 <: &{P} u32`).
-            let variance = std::mem::replace(&mut self.ambient_variance, ty::Variance::Covariant);
-
-            // Note: the order here is important. Create the placeholders first, otherwise
-            // we assign the wrong universe to the existential!
-            let b_replaced = self.instantiate_binder_with_placeholders(b);
-            let a_replaced = self.instantiate_binder_with_existentials(a);
-
-            self.relate(a_replaced, b_replaced)?;
-
-            self.ambient_variance = variance;
-        }
+        match self.ambient_variance {
+            ty::Variance::Covariant => {
+                // Covariance, so we want `for<..> A <: for<..> B` --
+                // therefore we compare any instantiation of A (i.e., A
+                // instantiated with existentials) against every
+                // instantiation of B (i.e., B instantiated with
+                // universals).
+
+                // Reset the ambient variance to covariant. This is needed
+                // to correctly handle cases like
+                //
+                //     for<'a> fn(&'a u32, &'a u32) == for<'b, 'c> fn(&'b u32, &'c u32)
+                //
+                // Somewhat surprisingly, these two types are actually
+                // **equal**, even though the one on the right looks more
+                // polymorphic. The reason is due to subtyping. To see it,
+                // consider that each function can call the other:
+                //
+                // - The left function can call the right with `'b` and
+                //   `'c` both equal to `'a`
+                //
+                // - The right function can call the left with `'a` set to
+                //   `{P}`, where P is the point in the CFG where the call
+                //   itself occurs. Note that `'b` and `'c` must both
+                //   include P. At the point, the call works because of
+                //   subtyping (i.e., `&'b u32 <: &{P} u32`).
+                let variance =
```

Comment:
I dont think there's any need to do this now? we already know the ambient variance is set to `Covariant`...